source_type.go

```diff
@@ -0,0 +1,46 @@
+package grpc_prometheus
+
+import (
+	"context"
+	"strconv"
+)
+
+type sourceContextKeyType struct{}
+
+// sourceContextKey unique key for all packages
+var sourceContextKey sourceContextKeyType
+
+type Source int
+
+const (
+	// Internal for requests that have header referer
+	Internal Source = 0
+
+	// External for other requests
+	External Source = 1
+)
+
+func (s Source) String() string {
+	switch s {
+	case Internal:
+		return "internal"
+	case External:
+		return "external"
+	default:
+		return "source(" + strconv.FormatInt(int64(s), 10) + ")"
+	}
+}
+
+// SetSourceToCtx returns derived context with source
+func SetSourceToCtx(ctx context.Context, source Source) context.Context {
+	return context.WithValue(ctx, sourceContextKey, source)
+}
+
+// GetSourceFromCtx returns source for given context
+func GetSourceFromCtx(ctx context.Context) Source {
+	v := ctx.Value(sourceContextKey)
+	if v == nil {
+		return Internal
```

Comment:
it is a default value, pls extract it to a `const` so that it is clear

Reply:
how about:
```go
func GetSourceFromCtx(ctx context.Context, defaultSource Source) (Source, bool) {
	v := ctx.Value(sourceContextKey)
	if v == nil {
		return defaultSource, false
	}
	return v.(Source), true
}
```

Reply:
it is not clear what is the 2nd return value provides. was default used or not?
if yes, when it can be useful/needed?

Reply:
So GetSourceFromCtx used inside https://github.com/Percona-Lab/go-grpc-prometheus and defaultSource cannot be override now so:
```go
func GetSourceFromCtx(ctx context.Context, defaultSource Source) (Source, bool) {
	v := ctx.Value(sourceContextKey)
	if v == nil {
		return defaultSource, false
	}
	return v.(Source), true
}
var defaultSource Source

func GetDefaultSource() (Source) {
	return defaultSource
}

func SetDefaultSource(newDefaultSource Source) {
	defaultSource = newDefaultSource
}

func GetSourceFromCtx(ctx context.Context, defaultSource Source) (Source) {
	v := ctx.Value(sourceContextKey)
	if v == nil {
		return defaultSource
	}
	return v.(Source)
}
```